merlin-server/sql/merlin/plan_collaboration.sql

```diff
@@ -0,0 +1,188 @@
+-- TODO list:
+--   - plan snapshot table (done)
+--   - duplicate function (done)
+--        - duplicate temporal subset of plan
+--   - merge request table
+--   - diff function
+--   - history tracking (tables and function) (done)
+--   - merge base function
+
+create table plan_snapshot_parent(
+  snapshot_id integer,
+  parent_snapshot_id integer
+);
+
+create table plan_latest_snapshot(
+  plan_id integer,
+  snapshot_id integer
+);
+
+-- Snapshot is a collection of the state of all the activities as they were at the time of the snapshot
+-- as well as any other properties of the plan that can change
+create table plan_snapshot(
+  snapshot_id integer generated always as identity,
+
+  plan_id integer not null,
+  revision integer not null,
+  name text not null,
+  duration interval not null,
+  start_time timestamptz not null
+);
+
+create table plan_snapshot_activities(
+  snapshot_id integer,
+  id integer,
+  name text,
+  tags text[],
+  source_scheduling_goal_id integer,
+  created_at timestamptz not null,
+  last_modified_at timestamptz not null,
+  start_offset interval not null,
+  type text not null,
+  arguments merlin_argument_set not null,
+  last_modified_arguments_at timestamptz not null,
+  metadata merlin_activity_directive_metadata_set
+);
+
+-- Captures the state of a plan and all of its activities
+create or replace function create_snapshot(plan_id integer)
+  returns integer -- snapshot id inserted into the table
+  language plpgsql as $$
+  declare
+    validate_planid integer;
+    inserted_snapshot_id integer;
+begin
+  select id from plan where plan.id = plan_id into validate_planid;
+  if validate_planid is null then
+    raise exception 'Plan % does not exist.', plan_id;
+  end if;
```

Comment:
this should be taken care of by foreign key constraints

Reply:
The check here and in `duplicate_plan` verify that the value of `plan_id` is not null, which is something that is not enforced by foreign key constraints. While I agree that if we add FKCs, this check can be simpler, we will still need to catch the null case here regardless.